toolchain/check/import_ref.cpp

```diff
@@ -224,6 +312,132 @@ class ImportRefResolver {
+    class_decl.class_id = context_.classes().Add({
+        .name_id = GetLocalNameId(import_class.name_id),
+        .enclosing_scope_id = NoEnclosingScopeForImports,
+        // `.self_type_id` depends on `class_id`, so is set below.
+        .self_type_id = SemIR::TypeId::Invalid,
+        .decl_id = class_decl_id,
+        .inheritance_kind = import_class.inheritance_kind,
+    });
+    // Write the function ID into the ClassDecl.
+    context_.ReplaceInstBeforeConstantUse(class_decl_id,
+                                          {Parse::NodeId::Invalid, class_decl});
+
+    // Build the `Self` type.
+    auto type_id = context_.GetClassType(class_decl.class_id);
+    context_.classes().Get(class_decl.class_id).self_type_id = type_id;
+
+    // Set a constant corresponding to the incomplete class.
+    // TODO: This uses the self type's constant, but the underlying instruction
+    // is NotConstant. Need to discuss whether we want to add a type_id to
+    // ClassDecl/InterfaceDecl and change behavior, which would also help
+    // consistency with other types. See also GetExprValueForLookupResult in
+    // handle_name.cpp.
+    auto const_id = context_.types().GetConstantId(type_id);
+    import_ir_constant_values_.Set(inst_id, const_id);
+    return const_id;
+  }
+
+  auto TryResolveTypedInst(SemIR::ClassDecl inst, SemIR::InstId inst_id,
+                           bool incomplete_type) -> SemIR::ConstantId {
+    const auto& import_class = import_ir_.classes().Get(inst.class_id);
+
+    SemIR::ConstantId const_id = SemIR::ConstantId::Invalid;
+    if (!incomplete_type) {
+      // If there's no incomplete type added, start by adding one for any
+      // recursive references.
+      const_id = CreateIncompleteClass(inst_id, import_class);
+      // If there's only a forward declaration, we're done.
+      if (!import_class.object_repr_id.is_valid()) {
+        return const_id;
+      }
+      // This may not be needed because all constants might be ready, but we do
+      // it here so that we don't need to track which work item corresponds to
+      // this instruction.
+      work_stack_.back().incomplete_type = true;
+    }
+
+    CARBON_CHECK(import_class.object_repr_id.is_valid())
+        << "Only reachable when there's a definition.";
+
+    // Load constants for the definition.
+    auto initial_work = work_stack_.size();
+
+    auto object_repr_const_id = GetLocalConstantId(import_class.object_repr_id);
+    auto base_const_id = import_class.base_id.is_valid()
+                             ? GetLocalConstantId(import_class.base_id)
+                             : SemIR::ConstantId::Invalid;
+
+    if (HasNewWork(initial_work)) {
+      return SemIR::ConstantId::Invalid;
+    }
+
+    // Update the class info with the definition.
+    if (incomplete_type) {
+      CARBON_CHECK(!const_id.is_valid());
+      const_id = import_ir_constant_values_.Get(inst_id);
+    }
+    auto& new_class = context_.classes().Get(
```

Comment:
Would it make sense to extract some of this logic into a helper function similar to `CreateIncompleteClass` and leave this to just wire things together?

Reply:
Done